### PR TITLE
Add fast-path for `rb_enc_from_index`.

### PR DESCRIPTION
**Pull request by nirvdrum**
This PR addresses a small problem I saw when benchmarking an application. I didn't trace it back far enough to see exactly which native extension was the root cause, but [fast_blank](https://rubygems.org/gems/fast_blank), [google-protobuf](https://rubygems.org/gems/google-protobuf), and [msgpack](https://rubygems.org/gems/msgpack) all make use of it.

This ties back into the issue where we can't store encodings by index in a simple array because users can replicate encodings. As such, we've put boundaries around any calls that look into the resulting `List` of `Encoding` objects. However, encoding replication is an uncommon feature and once an encoding has an index, it doesn't change for the remainder of the application. Additionally, most applications don't use more than three or four different encodings (e.g., `US_ASCII` for regexps, `UTF-8` for strings, and `ASCII_8BIT` for IO), so a cache is likely to handle the use cases of most applications.

### CODE REVIEW

File: src/main/java/org/truffleruby/core/encoding/EncodingNodes.java [hunk 1]
```diff
@@ -734,10 +734,22 @@ private RubyEncoding createDummy(String name) {
 
     @Primitive(name = "encoding_get_encoding_by_index", lowerFixnum = 0)
     public abstract static class GetEncodingObjectByIndexNode extends PrimitiveArrayArgumentsNode {
-        @Specialization
-        protected RubyEncoding getEncoding(int index) {
+
+        @Specialization(guards = "index == cachedIndex", limit = "getCacheLimit()")
```
Comment (eregon) [inline src/main/java/org/truffleruby/core/encoding/EncodingNodes.java:738]: This needs a `isSingleContext()` guard, as with multi-contexts, encoding indices could correspond to different encodings.

**Comment by eregon** [comment]
FWIW I'd like to remove the need to have a dynamic number of encodings in Ruby as notably here it adds the boundary instead of being a simple fixed-length array.
There is Encoding#replicate, which might not be too important (we could still make lookup by name work, but never actually give any String/Symbol/Regexp that encoding but the original one), but there is also `int rb_define_dummy_encoding(const char *name)`.
And pg specs seem to rely on being able to associate a dummy encoding with a Ruby String in https://github.com/ged/ruby-pg/blob/00cb2ecfaa70470234ee83efbd942b5e6ea0f4ea/spec/pg/connection_spec.rb#L1309.
I'm not sure how important that is, but it would be a breaking change.

It's still tempting though, giving the amount of complexity and cost just for these weird not-well-defined corner cases.

Something else we can do without breaking any compatibility would be to check if the index is  < `INITIAL_NUMBER_OF_ENCODINGS`, and then just look in a fixed-length RubyEncoding[] stored on RubyLanguage.

File: src/main/java/org/truffleruby/core/encoding/EncodingNodes.java (hunk 1 above)
Comment (nirvdrum) [reply src/main/java/org/truffleruby/core/encoding/EncodingNodes.java]: Thanks. I've added that and rebased the PR against the latest _master_.

**Comment by nirvdrum** [comment]
We originally had it as an `Encoding[]`, but changed it for compatibility. It might be a bit awkward, but maybe we can have a side data structure for the odd cases and use `Encoding[]` for all the defined, non-dummy encodings. Another thing perhaps worth looking at is lazy-loading the encodings, but retaining a fixed index. I'd have to dig through all the code again, but I'm pretty sure we eagerly load them in the native image to avoid issues with reflection. But, some of the encodings take a little while to load and can harm start-up time in the JVM.

**Comment by eregon** [comment]
> It might be a bit awkward, but maybe we can have a side data structure for the odd cases and use Encoding[] for all the defined, non-dummy encodings.

Yeah I think that makes sense, then for all initial encodings it would be an array read on a constant array (stored in RubyLanguage).
In that case I'm not sure how much the inline cache here would gain, but anyway let's merge this.

We could also maintain our own Encoding[] on the RubyContext (instead of an ArrayList), and just grow it by 1 every time a replicated/dummy encoding is added. The array itself would not be a constant then, and in multi-context we'd need to get the context to read it.

> Another thing perhaps worth looking at is lazy-loading the encodings, but retaining a fixed index. I'd have to dig through all the code again, but I'm pretty sure we eagerly load them in the native image to avoid issues with reflection. But, some of the encodings take a little while to load and can harm start-up time in the JVM.

I measured that (back then) and I couldn't see any difference in JVM startup. Encodings seem pretty fast to load, and loading them lazily just complicates things + adds extra checks. What's slow to load are transcoders due to the transcodings tables. Those tables are loaded eagerly only on SVM, and not on JVM.

**Review by eregon — APPROVED** [review]

**Pull request merged**